### PR TITLE
Kernel greedy option and Theory documentation 

### PR DESCRIPTION
**Pull request by efekhari27**
Hi Joseph, 

This PR is twofold:
- adding the theory page explaining the main methods and concepts;
- adding to KernelHerding and KernelHerdingTensorized an option ``is_greedy`` to the constructor. When passed True, this option multiplies the current potential by (m/m+1). 

I analyzed the effect of the greedy vs. herding criteria in the example/kernel_herding_vs_greedy.ipnb. On a uniform distribution, for a scale parameter set up by the heuristic of L.Pronzato, the herding seem to converge faster for higher dimensions. Then the empirically observed success of the support points might come from the energy-distance kernel.

Best, 
Elias

### CODE REVIEW

File: doc/theory/theory.rst [hunk 1]
```diff
@@ -1,21 +1,140 @@
...
+**Reproducing kernel Hilbert space.**
+
+Assuming that :math:`k` is a symmetric and positive definite function :math:`k: \cD_\vect{X} \times \cD_\vect{X} \rightarrow \Rset`, 
+latter called a "reproducing kernel" or simply a "kernel". A *reproducing kernel Hilbert space* (RKHS) is an inner product 
+space :math:`\cH(k)` of functions :math:`g:\cD_\vect{X} \rightarrow \Rset` with the following properties:
+
+* :math:`k(\cdot, \vect{x}) \in \cH(k), \quad \forall \vect{x} \in \cD_\vect{X}.`
+* :math:`\langle g, k(\cdot, \vect{x}) \rangle_{\cH(k)} = g(\vect{x}), \quad \forall \vect{x} \in \cD_\vect{X}, \forall g \in \cH(k)`
+
+Notice that for a defined reproducing kernel, a unique RKHS exists and vice versa (see `C.Oates, 2021 <https://arxiv.org/pdf/2109.06075.pdf>`_ ).
```
Comment (josephmure) [inline doc/theory/theory.rst:33]:
```suggestion
Note that for a defined reproducing kernel, a unique RKHS exists and vice versa (see `C.Oates, 2021 <https://arxiv.org/pdf/2109.06075.pdf>`_ ).
```
The original formulation implies that this is trivial.

File: doc/theory/theory.rst [hunk 2]
```diff
@@ -1,21 +1,140 @@
...
+
+Then, the potential of a discrete distribution :math:`\zeta_n = \frac1n \sum_{i=1}^{n} \delta(\vect{x}^{(i)})` 
+(uniform mixture of Dirac distributions at the design points :math:`\vect{X}_n`) associated with the kernel :math:`k` can be expressed as:
+
+.. math::
+    :name: design_potential
+    
+    P_{\zeta_n}(\vect{x}) = \frac1n \sum_{i=1}^{n} k(\vect{x}, \vect{x}^{(i)}).
+
+Close potentials can be interpreted to mean that the design $\bX_n$ adequately quantizes $\mu$
```
Comment (josephmure) [inline doc/theory/theory.rst:52]:
```suggestion
Close potentials can be interpreted to mean that the design :math:`\bX_n` adequately quantizes :math:`\mu`.
```

File: doc/theory/theory.rst [hunk 3]
```diff
@@ -1,21 +1,140 @@
...
+
+**Maximum mean discrepancy.**
+
+A recent metric of discrepancy and quadrature error is offered by the MMD.
```
Comment (josephmure) [inline doc/theory/theory.rst:56]: Is it recent?

File: doc/theory/theory.rst [hunk 4]
```diff
@@ -1,21 +1,140 @@
...
+    
+    P_{\zeta_n}(\vect{x}) = \frac1n \sum_{i=1}^{n} k(\vect{x}, \vect{x}^{(i)}).
+
+Close potentials can be interpreted to mean that the design $\bX_n$ adequately quantizes $\mu$
+
+**Maximum mean discrepancy.**
+
+A recent metric of discrepancy and quadrature error is offered by the MMD. 
+This distance between two distributions :math:`\mu` and :math:`\zeta` is given by the 
+maximal quadrature error committed for any function within a defined RKHS:
```
Comment (josephmure) [inline doc/theory/theory.rst:58]:
```suggestion
maximal quadrature error committed for any function within the unit ball of an RKHS:
```

File: doc/theory/theory.rst [hunk 5]
```diff
@@ -1,21 +1,140 @@
...
+    :name: mmd
+    
+    \mathrm{MMD}_k(\mu, \zeta) := 
+    \sup_{\lVert g \lVert_{\cH(k)} \leq 1}
+            \left | \int_{\cD_{\vect{X}}} g(\vect{x}) \di \mu(\vect{x}) - \int_{\cD_{\vect{X}}} g(\vect{x}) \di \zeta(\vect{x}) \right|.
+
+Using the Cauchy-Schwartz inequality, one can demonstrate that 
+:math:`\mathrm{MMD}_k(\mu, \zeta) = \left\lVert P_{\mu}(\vect{x}) - P_{\zeta}(\vect{x}) \right\lVert_{\cH(k)}` 
+(see `C.Oates, 2021 <https://arxiv.org/pdf/2109.06075.pdf>`_ ). 
+Moreover, kernels are called characteristic when the following implication is true , :math:`\mathrm{MMD}_k(\mu, \zeta) = 0 \Rightarrow \mu = \zeta`.
```
Comment (josephmure) [inline doc/theory/theory.rst:70]:
```suggestion
Moreover, a kernel :math:`k` is called characteristic if :math:`\mathrm{MMD}_k(\mu, \zeta) = 0` is equivalent to :math:`\mu = \zeta`.
```

File: doc/theory/theory.rst [hunk 6]
```diff
@@ -1,21 +1,140 @@
...
+
+.. math::
+  :name: kh_estimation
+
+    \vect{x}^{(n+1)} \in \argmin_{\vect{x}\in\mathcal{S}} \left( \frac{1}{n} \sum_{i=1}^{n} k(\vect{x},\vect{x}^{(i)}) 
+    - \frac{1}{N} \sum_{j=1}^N k(\vect{x},\vect{x}'^{(j)}) \right) \,.
+
+When no observation is available, which is the common situation at the design stage, 
+the kernel hyperparameters (e.g., correlation lengths) have to be set to heuristic values. 
+MMD minimization is quite versatile and was explored in more details by (Teymur & al., 2021)
```
Comment (josephmure) [inline doc/theory/theory.rst:104]:
```suggestion
MMD minimization is quite versatile and was explored in detail by (Teymur & al., 2021)
```
Not sure about this one. What do you think?

File: doc/theory/theory.rst [hunk 7]
```diff
@@ -1,21 +1,140 @@
...
+
+.. math::
+    :name: greedy_criterion
+
+    \vect{x}^{(n+1)} \in \argmin_{\vect{x}\in\mathcal{S}} \Bigg( \frac{1}{N} \sum_{j=1}^N \|\vect{x}-\vect{x}'^{(j)}\| 
+    - \frac{1}{n+1} \sum_{i=1}^{n} \|\vect{x}-\vect{x}^{(i)}\| \Bigg) \,.
+
+For this criterion, one can notice that it is almost identical to the KH one in :ref:`(4) <kh_criterion>` when 
+taking as kernel the energy-distance kernel given in :ref:`(6) <energy_kernel>`.
+These two iterative methods were exploited in (Fekhari & al., 2021) to study new ways to construct
```
Comment (josephmure) [inline doc/theory/theory.rst:135]: Maybe add link to PDFs of articles when possible?

File: otkerneldesign/KernelHerding.py [hunk 8]
```diff
@@ -33,6 +33,10 @@ class KernelHerding:
         If not specified, then *distribution* and *candidate_set_size* must be in order to generate it automatically.
     initial_design : 2-d list of float
         Sample of points that must be included in the design. Empty by default.
+    is_greedy : Boolean
```
Comment (josephmure) [inline otkerneldesign/KernelHerding.py:36]: I would advise dropping this option, as your experiments show it does not change much.

**Review by josephmure — COMMENTED** [review]
Good job! A few things need to be polished a little before publication IMO.

**Comment by efekhari27** [comment]
Hi Joseph, 
Thanks a lot for your detailed review. I took into account your remarks and worked on a new example to illustrate the use of the package in a validation context. 
I tag this PR as the 0.1.4 version of the package that I will update shortly into PyPI. 
Elias

**Pull request merged**